### ISSUE TITLE
cache.db-wal gets really big over time

### ISSUE THREAD

**Issue by trou** [LMS-Community/slimserver#179]
Hello,
I've hit a problem several times with the SQLite engine: the `cache.db-wal` [write ahead log](https://www.sqlite.org/wal.html), gets really big (several GB) after some time.

Maybe it would be nice to either disable WAL or use the [pragma](https://sqlite.org/pragma.html) `PRAGMA schema.journal_size_limit` to avoid such problems ?

Version: 
`Logitech Media Server Version: 7.9.1 - 1489743085 @ Fri Mar 17 09:42:01 UTC 2017`

running on Debian stable, amd64

**Comment by mherger** [comment]
Several GB certainly isn't normal. How large is your collection? Are you using any 3rd party plugins? How often do you restart your LMS?

**Comment by trou** [comment]
My library isn't that big: `Total Tracks: 19,162`.

3rd party plugins:
* Network Test
* Song Scanner

And I basically never restart it, usually it can run for months... but I just added a crontab to restart it more often to workaround that problem.

**Comment by mherger** [comment]
Depending on when your cache.db file was created it wouldn't auto_vacuum yet. You might want to delete it once, as this setting can only be set when the database is created.

BTW: the cache.db has nothing to do with the library size. Sorry for the confusion. It's more used when using online services, internet radio etc. But the file should not grow to that size.

**Comment by trou** [comment]
Since it has happened a few times, I already deleted it in the past and it didn't change anything apparently.

I use podcasts and live365 quite often. I think it may be related to podcasts use.

**Comment by mherger** [comment]
As you seem to be familiar with SQLite, could you please check what `pragma auto_vacuum;` returns? What about `select count(1) from cache`?

**Comment by trou** [comment]
Note that I restarted the server 2 days ago:
```
# ls -lh cache.db*
-rw-r--r-- 1 squeezeboxserver nogroup  26M déc.   6 01:36 cache.db
-rw-r--r-- 1 squeezeboxserver nogroup 768K déc.   6 09:01 cache.db-shm
-rw-r--r-- 1 squeezeboxserver nogroup  97M déc.   6 09:00 cache.db-wal

# /var/lib/squeezeboxserver/cache# sqlite3 cache.db
SQLite version 3.16.2 2017-01-06 16:32:41
Enter ".help" for usage hints.
sqlite> pragma auto_vacuum;
1
sqlite> select count(1) from cache;
11776
sqlite> select k,length(v),t from cache order by length(v) desc limit 10;
4164457312498764|5476196|1512690232
34440204939271928|5476196|1512690180
2236923668156905971|600258|1512951759
106655874568059686|341407|1512690156
287229222134527995|341407|1513123955
338573358764984921|341407|1513123955
438034030229150162|341407|1512690156
986232649147533362|341407|1513123959
1308760823120954843|341407|1513037773
1953965665245828399|341407|1513037773
```

**Comment by mherger** [comment]
A few more queries:
* get the timestamp of the last cleanup: `select * from cache where k = 955705662110008944;`
* get oldest items in the cache: `select t, length(v) from cache where t > -1 group by t order by t limit 10;`

Use something like https://www.gaijin.at/olsutc.php to convert the timestamps. When was the last cleanup? What's the expiry time (t) for the oldest entries? Their size?

I must say that you have a lot more entries in the cache than I do in an environment where we do a lot of Spotify and internet radio listening. The plugins you mentioned aren't really 3rd party. Would you have the full list of plugins?

**Comment by trou** [comment]
```
sqlite> select * from cache where k = 955705662110008944;
955705662110008944|1512529821|-1
sqlite> select t, length(v) from cache where t > -1 group by t order by t limit 10;
1512550500|48
1512550501|66
1512550576|48
1512551712|106915
1512554449|48
1512637812|109470
1512690156|341407
1512690157|341407
1512690180|5476196
1512690232|5476196
```
```
irb(main):004:0> Time.at(1512529821)
=> 2017-12-06 04:10:21 +0100
irb(main):005:0> Time.at(1512550500)
=> 2017-12-06 09:55:00 +0100
```
here is the plugin list:
![image](https://user-images.githubusercontent.com/1587988/33655767-4ad89082-da74-11e7-8285-9cf5663f5d42.png)

**Comment by mherger** [comment]
Thanks! Unfortunately I got the second query wrong. I wanted to see the items with the longest expiry: `select t, length(v) from cache where t > -1 group by t order by t desc limit 10;`.

Considering your small list of plugins, I'm wondering where those largish entries come from. Are you using many different podcasts? Some podcasts with huge lists of items? Could you send me your podcasts.prefs file (michael ät slimdevices dot com)?

**Comment by mherger** [comment]
I'm running out of ideas here. Please keep monitoring the file size. It's currently still reasonable. But please let me know should you see it grow beyond the GB.

**Comment by trou** [comment]
What about using the pragma to limit the size and force the WAL to flush ?

**Comment by mherger** [comment]
I'm sorry, totally missed that point. Did you try the journal_size_limit pragma? As I don't see the growth you're seeing, you might be in a good position to give it a try. I've added it to my collection, but will have to wait a day or two to see whether it's being picked up.

**Comment by trou** [comment]
Apparently pragmas are linked to the connection and not to the database itself. So it would have to be added to Slimserver's code.

I did try however to run `PRAGMA wal_checkpoint(TRUNCATE)` and it effectively merged the wal and reset it to zero.

**Comment by mherger** [comment]
Try something like

```diff --git a/Slim/Utils/DbCache.pm b/Slim/Utils/DbCache.pm
index 0d17b73b4..816ba051e 100644
--- a/Slim/Utils/DbCache.pm
+++ b/Slim/Utils/DbCache.pm
@@ -237,6 +237,7 @@ sub _init_db {
 		$dbh->do('PRAGMA journal_mode = WAL');
 		# scanner is heavy on writes, server on reads - tweak accordingly
 		$dbh->do('PRAGMA wal_autocheckpoint = ' . (main::SCANNER ? 10000 : 200));
+		$dbh->do('PRAGMA journal_size_limit = 100000000');
 
 		my $dbhighmem;
 		if (main::RESIZER) {
```
You might want to play with the value, of course.

**Comment by mherger** [comment]
One more thing: you don't have other processes connect to that database by chance? See https://www.sqlite.org/wal.html#avoiding_excessively_large_wal_files

**Comment by trou** [comment]
Not that I know of.

**Comment by mherger** [comment]
My patch actually seems to be working just fine: I set it to 2MB on my main system. While I saw it grow beyond 5MB at times, it now settled at exactly the value I defined. That's expected behaviour: while processing is going on, SQLite would take whatever amount of disk space it needs. The journal_size_limit is only there to tell it the maximum size to leave sitting around available to be re-used.

I'll try to implement this soon. Thanks for the heads up!
